students/tammyd/lesson10/mailroom2_working.py

```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python
+"""
+mailroom assignment
+This version uses a dict for the main db, and exception handling to
+check input, and has been factored to be amenable to testing.
+"""
+
+import sys
+import math
+
+from functools import reduce
+
+# handy utility to make pretty printing easier
+from textwrap import dedent
+
+
+# In memory representation of the donor database
+# using a tuple for each donor
+# -- kind of like a record in a database table
+# using a dict with a lower case version of the donor's name as the key
+# This makes it easier to have a 'normalized' key.
+#  you could get a bit fancier by having each "record" be a dict, with
+#   "name" and "donations" as keys.
+def get_donor_db():
+    return {'william gates iii': ("William Gates III", [653772.32, 12.17]),
+            'jeff bezos': ("Jeff Bezos", [877.33]),
+            'paul allen': ("Paul Allen", [663.23, 43.87, 1.32]),
+            'mark zuckerberg': ("Mark Zuckerberg", [1663.23, 4300.87, 10432.0]),
+            }
+
+class Donor:
+    def __init__(self, name, donations=None):
+        self.name = name
+        self._donations = [] if donations is None else donations
+
+    @property
+    def name(self):
+        return self._name
+
+    @property
+    def donations(self):
+        return self._donations
+
+    def add_donations(self, donations):
+        self._donations.append(donation)
+
+    @property
+    def total_donations(self):
+        return reduce(lambda a,x: a+t, self._donations, 0)
+        #avoid the style below
+        # s = 0
+        # for d in self._donations:
+        #     s += n 
+        # return s
+
+    def __str__(self):
+        return "{}: {}".format(self._name, self._donations)
+
+class DonorDB:
+    def __init__(self):
+        self._donors = {}
+
+    def add_donor(sef, donor):
+        self._donors[donor.name.lower()] = donor
+
+    def get_total_from_donor(self, donor_name):
+        self._donors[donor_name.lower()].total_donations
+
+    def get_donor(self, donor_name):
+        return self._donors[donor_name.lower()] ###stuff here
+
+    @property
+    def num_donors(self):
+        return str(self._donors)
+
+    def __repr__(self):
+        return #whaaaax
+
+def get_donor_db_oo():
+    db = DonorDB()
+    raw_data = get_donor_db()
+
+    for k, v in raw_data.items():
+        donor = Donor(k)
```

Comment:
Donor can take alist of donations to initialize, so this can be done in one fell swoop ;-)